File: intact_agent_runner/commands.py
from __future__ import annotations

from dataclasses import dataclass
import os
import subprocess
from typing import Mapping


@dataclass
class CommandResult:
    command: str
    cwd: str | None
    code: int | None
    signal: str | None
    timed_out: bool
    stdout: str
    stderr: str

    @property
    def ok(self) -> bool:
        return self.code == 0 and not self.timed_out


def run_command(
    command: str,
    args: list[str],
    *,
    cwd: str | None = None,
    env: Mapping[str, str] | None = None,
    input_text: str | None = None,
    timeout_ms: int = 120000,
) -> CommandResult:
    merged_env = os.environ.copy()
    if env:
        merged_env.update(env)
    command_text = " ".join([command, *args])
    try:
        completed = subprocess.run(
            [command, *args],
            cwd=cwd,
            env=merged_env,
            input=input_text,
            text=True,
            capture_output=True,
            timeout=timeout_ms / 1000,
            check=False,
        )
        return CommandResult(
            command=command_text,
            cwd=cwd,
            code=completed.returncode,
            signal=None,
            timed_out=False,
            stdout=completed.stdout,
            stderr=completed.stderr,
        )
    except subprocess.TimeoutExpired as error:
        return CommandResult(
            command=command_text,
            cwd=cwd,
            code=None,
            signal="SIGTERM",
            timed_out=True,
            stdout=error.stdout or "",
            stderr=error.stderr or "",
        )


def git_status(repo_root: str) -> CommandResult:
    return run_command("git", ["status", "--short", "--branch"], cwd=repo_root, timeout_ms=30000)


def command_summary(result: CommandResult) -> str:
    parts = [f"$ {result.command}"]
    if result.cwd:
        parts.append(f"cwd: {result.cwd}")
    parts.append(f"exit: {'timeout' if result.timed_out else result.code}")
    if result.stdout.strip():
        parts.append(f"stdout:\n{result.stdout.strip()}")
    if result.stderr.strip():
        parts.append(f"stderr:\n{result.stderr.strip()}")
    return "\n".join(parts)

File: src/commands.js
import { spawn } from "node:child_process";


export function runCommand(command, args, {
  cwd,
  env = {},
  input = undefined,
  timeoutMs = 120000,
} = {}) {
  return new Promise((resolve) => {
    const child = spawn(command, args, {
      cwd,
      env: { ...process.env, ...env },
      stdio: ["pipe", "pipe", "pipe"],
    });
    let stdout = "";
    let stderr = "";
    let timedOut = false;
    const timer = setTimeout(() => {
      timedOut = true;
      child.kill("SIGTERM");
    }, timeoutMs);

    child.stdout.on("data", (chunk) => { stdout += chunk; });
    child.stderr.on("data", (chunk) => { stderr += chunk; });
    child.on("close", (code, signal) => {
      clearTimeout(timer);
      resolve({
        command: [command, ...args].join(" "),
        cwd,
        code,
        signal,
        timedOut,
        stdout,
        stderr,
        ok: code === 0 && !timedOut,
      });
    });

    if (input) child.stdin.write(input);
    child.stdin.end();
  });
}

export async function gitStatus(repoRoot) {
  const result = await runCommand("git", ["status", "--short", "--branch"], {
    cwd: repoRoot,
    timeoutMs: 30000,
  });
  return result;
}

export function commandSummary(result) {
  const parts = [`$ ${result.command}`];
  if (result.cwd) parts.push(`cwd: ${result.cwd}`);
  parts.push(`exit: ${result.timedOut ? "timeout" : result.code}`);
  if (result.stdout.trim()) parts.push(`stdout:\n${result.stdout.trim()}`);
  if (result.stderr.trim()) parts.push(`stderr:\n${result.stderr.trim()}`);
  return parts.join("\n");
}
